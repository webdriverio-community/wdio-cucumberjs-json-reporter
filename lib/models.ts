import { Argument, Tag } from '@wdio/reporter';


export interface BrowserData {
    browser: {
        name: string;
        version: string;
    };
}

export interface AppData{
    app?: {
        name: string;
        version: string;
    };
}

export interface MetadataObject {
    app?: {
        name: string;
        version: string;
    };
    browser?: {
        name: string;
        version: string;
    };
    device?: string;
    platform?: {
        name: string;
        version: string;
    };
    keyword?: string;
    type?: string;
    foo?: string;
}


export interface Report {
    feature: Feature;
}

export interface FeatureMetadata {
    keyword: string;
    type: string;
}

export interface Feature {
    report?: {
        keyword: string;
    };
    keyword?: string;
    type?: string;
    metadata?: MetadataObject;
    description?: string;
    line?: number | null;
    name?: string;
    uri?: string;
    tags?: string[] | Tag[] | string;
    elements?: Scenario[];
    id?: string;
}

export interface Scenario {
    keyword?: string;
    type?: string;
    description?: string;
    name?: string;
    tags?: string[] | Tag[] | string;
    id?: string;
    steps?: Step[];
    foo?: string;
    bar?: string;
    foobar?: string;
}

export interface Step {
    arguments?: ( string | Argument )[];
    keyword?: string;
    name?: string;
    result?: {
        status: string;
        duration: number;
        error_message?: string;
    };
    line?: number | null;
    match?: {
        location: number | string;
    };
    embeddings?: Embedding[];
    foo?: string;
}

export interface Result extends ErrorMessage {
    status: string;
    duration: number;
}

export interface CucumberJsAttachment {
    data: string;
    type: string;
}

export interface Embedding {
    data: string;
    mime_type: string;
}

export interface ErrorMessage {
    error_message?: string;
}

export interface cjson_metadata {
    browser?: {
        name: string;
        version: string;
    };
    device?: string;
    app?: {
        name: string;
        version: string;
    };
    platform?: {
        name: string;
        version: string;
    };
}
