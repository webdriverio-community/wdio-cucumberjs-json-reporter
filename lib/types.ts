import { Reporters } from '@wdio/types';

export interface CucumberJsJsonReporterInterface extends Reporters.Options {
    jsonFolder?: string;
    language?: string;
}
