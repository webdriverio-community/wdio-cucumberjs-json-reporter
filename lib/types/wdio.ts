import { Capabilities, Options } from '@wdio/types';
import WDIOReporterBaseOptions, { Argument, HookStats, RunnerStats, SuiteStats, Tag, TestStats } from '@wdio/reporter';
import { WriteStream } from 'fs';
import { cjson_metadata } from '../models';

export interface W3CCapabilitiesExtended extends Capabilities.W3CCapabilities {
    cjson_metadata?: cjson_metadata;
    testobject_app_id?: string;
    app?: string;
}

export interface DesiredCapabilitiesExtended extends Capabilities.DesiredCapabilities {
    cjson_metadata?: cjson_metadata;
    testobject_app_id?: string;
    app?: string;
}

export interface TestrunnerExtended extends Options.Testrunner {
    jsonFolder?: string;
    language?: string;
    cjson_metadata?: cjson_metadata;
    logFile?: string;
    stdout?: boolean;
    writeStream?: WriteStream;
}
export interface WebdriverIOExtended extends Options.WebdriverIO {
    app?: string;
    requestedCapabilities?: {
        cjson_metadata?: cjson_metadata;
        w3cCaps: {
            alwaysMatch?: {
                foo?: true;
            };
        };
    };
}

export interface WDIOReporterBaseOptionsExtended extends WDIOReporterBaseOptions {
    jsonFolder: string;
    language: string;
}

export interface HookStatsExtended extends HookStats {
    description?: string;
    tags?: string[] | Tag[] | string;
    keyword?: string;
    argument?: Argument;
    line?: number;
    name?: string;
}

export class TestStatsExtended extends TestStats {
    public description?: string;
    public tags?: string[] | Tag[] | string;
    public keyword?: string;
    public name?: string;
    public id?: string;
    public foo?: string | boolean;
    public bar?: boolean;
}

export interface RunnerStatsExtended extends RunnerStats {
    metadata?: {
        foo?: string;
    };
}

export interface SuiteStatsExtended extends SuiteStats {
    keyword?: string;
    line?: number;
    name?: string;
    id?: string;
}
