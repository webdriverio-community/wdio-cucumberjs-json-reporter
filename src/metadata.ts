import { browser } from '@wdio/globals'

import { NOT_KNOWN } from './constants.js'
import type { AppData, BrowserData, MetadataObject, cjson_metadata } from './types'
import type {
    DesiredCapabilitiesExtended,
    RunnerStatsExtended,
    W3CCapabilitiesExtended,
    WebdriverIOExtended,
} from './types/wdio'

export class Metadata {
    /**
     * Determine the metadata that needs to be added
     */
    public determineMetadata(data: RunnerStatsExtended): MetadataObject {
        let instanceData: AppData | BrowserData
        const currentCapabilities = data.capabilities as WebdriverIO.Capabilities
        const optsCaps = browser.requestedCapabilities as W3CCapabilitiesExtended
        const currentConfigCapabilities = data?.capabilities as DesiredCapabilitiesExtended
        const w3cCaps: cjson_metadata = (data.capabilities)
            ? Object.prototype.hasOwnProperty.call(data.capabilities, 'cjson:metadata')
                // Fixes: https://github.com/webdriverio-community/wdio-cucumberjs-json-reporter/issues/73
                ? data.capabilities['cjson:metadata'] as cjson_metadata
                // Fallback
                : (browser as WebdriverIOExtended)?.requestedCapabilities?.cjson_metadata
            : {}
        const metadata: cjson_metadata = (currentConfigCapabilities as W3CCapabilitiesExtended)?.cjson_metadata
            || w3cCaps // When an app is used to test
            || (optsCaps as DesiredCapabilitiesExtended)?.cjson_metadata // devtools
            || {} as cjson_metadata

        // When an app is used to test
        if (currentConfigCapabilities?.app || metadata?.app) {
            instanceData = this.determineAppData(currentConfigCapabilities, metadata)
        } else {
            // Then a browser
            instanceData = this.determineBrowserData(currentCapabilities, metadata)
        }

        return <MetadataObject>{
            ...instanceData,
            device: this.determineDeviceName(metadata, currentConfigCapabilities),
            platform: {
                name: this.determinePlatformName(metadata, currentCapabilities),
                version: this.determinePlatformVersion(metadata, currentCapabilities),
            },
        }
    }

    /**
     * Determine the device name
     */
    public determineDeviceName(metadata: cjson_metadata, currentConfigCapabilities: DesiredCapabilitiesExtended): string {
        return (metadata?.device || currentConfigCapabilities?.['cjson:metadata']?.device || `Device name ${NOT_KNOWN}`)
    }

    /**
     * Determine the platform name
     */
    public determinePlatformName(metadata: cjson_metadata, currentCapabilities: WebdriverIO.Capabilities): string {
        const currentPlatformName = currentCapabilities?.platformName
            ? currentCapabilities?.platformName.includes('mac')
                ? 'osx'
                : currentCapabilities.platformName.includes('windows')
                    ? 'windows'
                    : currentCapabilities?.platformName
            : `Platform name ${NOT_KNOWN}`
        return (metadata.platform && metadata?.platform?.name)
            ? metadata.platform?.name
            : currentPlatformName
    }

    /**
     * Determine the platform version
     */
    public determinePlatformVersion(metadata: cjson_metadata, currentCapabilities?: WebdriverIO.Capabilities): string {
        if (metadata && metadata.platform && metadata.platform?.version) {
            return metadata.platform.version
        }
        if (currentCapabilities?.['appium:platformVersion']) {
            return currentCapabilities?.['appium:platformVersion']
        }
        return `Version ${NOT_KNOWN}`
    }

    /**
     * Determine the app data
     */
    public determineAppData(currentConfigCapabilities: DesiredCapabilitiesExtended, metadata: cjson_metadata): AppData {
        const metaAppName: string = (metadata?.app && metadata.app?.name) ? metadata?.app?.name : 'No metadata.app.name available'
        const metaAppVersion: string = (metadata?.app && metadata.app.version) ? metadata.app.version : 'No metadata.app.version available'
        const appPath = currentConfigCapabilities.app || metaAppName
        const appName = appPath.substring(appPath.replace('\\', '/').lastIndexOf('/')).replace('/', '')

        return {
            app: {
                name: appName,
                version: metaAppVersion,
            },
        }
    }

    /**
     * Determine the browser data
     */
    public determineBrowserData(currentCapabilities: WebdriverIO.Capabilities, metadata: cjson_metadata): BrowserData {
        const browserName = currentCapabilities?.browserName
            || ((metadata && metadata?.browser && metadata.browser?.name) ? metadata?.browser?.name : 'No metadata.browser.name available')
        const browserVersion = currentCapabilities?.browserVersion
            || ((metadata && metadata?.browser && metadata?.browser?.version) ? metadata?.browser?.version : 'No metadata.browser.version available')

        return <BrowserData>{
            browser: {
                name: browserName,
                version: browserVersion,
            }
        }
    }
}
