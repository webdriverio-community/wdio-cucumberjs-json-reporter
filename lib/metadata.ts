import { AppData, BrowserData, MetadataObject, cjson_metadata } from './models';
import { DesiredCapabilitiesExtended, RunnerStatsExtended, W3CCapabilitiesExtended, WebdriverIOExtended } from './types/wdio';
import { Browser } from 'webdriverio';
import { NOT_KNOWN } from './constants';
import WebDriver from 'webdriver';

export class Metadata {
    /**
     * Determine the metadata that needs to be added
     */
    public determineMetadata ( data: RunnerStatsExtended ): MetadataObject {
        let instanceData: AppData | BrowserData;
        const currentCapabilities = data.capabilities as W3CCapabilitiesExtended;
        const optsCaps = ( browser as Browser<'async'> ).options.capabilities as W3CCapabilitiesExtended;
        const currentConfigCapabilities = data?.capabilities as DesiredCapabilitiesExtended;
        const w3cCaps: cjson_metadata = Object.prototype.hasOwnProperty.call( data.config.capabilities, 'cjson:metadata' )
            // Fixes: https://github.com/webdriverio-community/wdio-cucumberjs-json-reporter/issues/73
            ? data.config.capabilities['cjson:metadata'] as cjson_metadata
            // Fallback
            : ( ( browser as Browser<'async'> ).options as WebdriverIOExtended )?.requestedCapabilities?.cjson_metadata;
        const metadata: cjson_metadata = ( currentConfigCapabilities as W3CCapabilitiesExtended )?.cjson_metadata
            || w3cCaps // When an app is used to test
            || ( optsCaps as DesiredCapabilitiesExtended )?.cjson_metadata // devtools
            || {} as cjson_metadata;

        // When an app is used to test
        // eslint-disable-next-line @typescript-eslint/tslint/config
        if ( currentConfigCapabilities?.app || metadata?.app ) {
            instanceData = this.determineAppData( currentConfigCapabilities, metadata );
        } else {
            // Then a browser
            instanceData = this.determineBrowserData( currentCapabilities, currentConfigCapabilities, metadata );
        }

        return <MetadataObject>{
            ...instanceData,
            device: this.determineDeviceName( metadata, currentConfigCapabilities ),
            platform: {
                name: this.determinePlatformName( metadata, currentCapabilities ),
                version: this.determinePlatformVersion( metadata ),
            },
        };
    }

    /**
     * Determine the device name
     */
    public determineDeviceName ( metadata: cjson_metadata, currentConfigCapabilities: WebDriver.DesiredCapabilities ): string {
        return ( metadata?.device || currentConfigCapabilities?.deviceName || `Device name ${NOT_KNOWN}` );
    }

    /**
     * Determine the platform name
     */
    public determinePlatformName ( metadata: cjson_metadata, currentCapabilities: WebDriver.DesiredCapabilities ): string {
        const currentPlatformName = currentCapabilities?.platformName
            ? currentCapabilities?.platformName.includes( 'mac' )
                ? 'osx'
                : currentCapabilities.platformName.includes( 'windows' )
                    ? 'windows'
                    : currentCapabilities?.platformName
            : `Platform name ${NOT_KNOWN}`;
        return ( metadata.platform && metadata?.platform?.name )
            ? metadata.platform?.name
            : currentPlatformName;
    }

    /**
     * Determine the platform version
     */
    public determinePlatformVersion ( metadata: cjson_metadata ): string {
        return ( metadata && metadata.platform && metadata.platform?.version )
            ? metadata.platform.version
            : `Version ${NOT_KNOWN}`;
    }

    /**
     * Determine the app data
     */
    public determineAppData ( currentConfigCapabilities: DesiredCapabilitiesExtended, metadata: cjson_metadata ): AppData {
        const metaAppName: string = ( metadata?.app && metadata.app?.name ) ? metadata?.app?.name : 'No metadata.app.name available';
        const metaAppVersion: string = ( metadata?.app && metadata.app.version ) ? metadata.app.version : 'No metadata.app.version available';
        const appPath = currentConfigCapabilities.app || metaAppName;
        const appName = appPath.substring( appPath.replace( '\\', '/' ).lastIndexOf( '/' ) ).replace( '/', '' );

        return {
            app: {
                name: appName,
                version: metaAppVersion,
            },
        };
    }

    /**
     * Determine the browser data
     */
    public determineBrowserData ( currentCapabilities: WebDriver.DesiredCapabilities, currentConfigCapabilities: WebDriver.DesiredCapabilities, metadata: cjson_metadata ): BrowserData {
        const browserName = currentCapabilities?.browserName
            || currentConfigCapabilities?.browserName
            || ( ( metadata && metadata?.browser && metadata.browser?.name ) ? metadata?.browser?.name : 'No metadata.browser.name available' );
        const browserVersion = currentCapabilities?.version
            || currentCapabilities?.browserVersion
            || currentConfigCapabilities?.browserVersion
            || ( ( metadata && metadata?.browser && metadata?.browser?.version ) ? metadata?.browser?.version : 'No metadata.browser.version available' );

        return <BrowserData>{
            browser: {
                name: browserName,
                version: browserVersion,
            }
        };
    }
}
