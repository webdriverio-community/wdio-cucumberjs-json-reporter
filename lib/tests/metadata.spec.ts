import { CAPS_METADATA_RUNNER_STATS, FULL_RUNNER_STATS, SMALL_RUNNER_STATS, WDIO6_RUNNER_STATS } from './__mocks__/mocks';
import { W3CCapabilitiesExtended, WebdriverIOExtended } from '../types/wdio';
import { Browser } from 'webdriverio';
import { Metadata } from '../metadata';
import { NOT_KNOWN } from '../constants';
import { Options } from '@wdio/types';
import WebDriver from 'webdriver';
import { cjson_metadata } from '../models';

describe( 'metadata', () => {
    let metadataClassObject: Metadata;
    beforeAll( () => {
        metadataClassObject = new Metadata();
    } );

    describe( 'determineAppData', () => {
        it( 'should return that no app metadata could be determined', () => {
            expect( metadataClassObject.determineAppData(
                {} as WebDriver.DesiredCapabilities,
                {} as cjson_metadata,
            ) ).toMatchSnapshot();
        } );

        it( 'should return that the app name and version based on the metadata', () => {
            expect( metadataClassObject.determineAppData(
                {} as WebDriver.DesiredCapabilities,
                {
                    app: {
                        name: 'metadata app name',
                        version: 'metadata version 1.2.3',
                    },
                } as cjson_metadata,
            ) ).toMatchSnapshot();
        } );

        it( 'should return that the app name based on the config.app', () => {
            expect( metadataClassObject.determineAppData(
                {
                    app: 'here/there/app.apk',
                } as WebDriver.DesiredCapabilities,
                {} as cjson_metadata,
            ) ).toMatchSnapshot();
        } );
    } );

    describe( 'determineBrowserData', () => {
        it( 'should return that no browser metadata could be determined', () => {
            expect( metadataClassObject.determineBrowserData(
                {} as WebDriver.DesiredCapabilities,
                {} as WebDriver.DesiredCapabilities,
                {} as cjson_metadata,
            ) ).toMatchSnapshot();
        } );

        it( 'should return that the browser name and version based on the metadata', () => {
            expect( metadataClassObject.determineBrowserData(
                {} as WebDriver.DesiredCapabilities,
                {} as WebDriver.DesiredCapabilities,
                {
                    browser: {
                        name: 'metadata browser name',
                        version: 'metadata version 1.2.3',
                    },
                } as cjson_metadata,
            ) ).toMatchSnapshot();
        } );

        it( 'should return that the browser name and version based on the capabilities when version is used', () => {
            expect( metadataClassObject.determineBrowserData(
                {
                    browserName: 'capabilities browser name',
                    version: 'capabilities version',
                } as WebDriver.DesiredCapabilities,
                {} as WebDriver.DesiredCapabilities,
                {} as cjson_metadata,
            ) ).toMatchSnapshot();
        } );

        it( 'should return that the browser name and version based on the capabilities when browserVersion is used', () => {
            expect( metadataClassObject.determineBrowserData(
                {
                    browserName: 'capabilities browser name',
                    browserVersion: 'capabilities browserVersion',
                } as WebDriver.DesiredCapabilities,
                {} as WebDriver.DesiredCapabilities,
                {} as cjson_metadata,
            ) ).toMatchSnapshot();
        } );

        it( 'should return that the browser name and version based on the configCapabilities', () => {
            expect( metadataClassObject.determineBrowserData(
                {} as WebDriver.DesiredCapabilities,
                {
                    browserName: 'configCapabilities browser name',
                    browserVersion: 'configCapabilities browser version',
                } as WebDriver.DesiredCapabilities,
                {} as cjson_metadata,
            ) ).toMatchSnapshot();
        } );
    } );

    describe( 'determineDeviceName', () => {
        it( 'should be able to return the device metadata based on the metadata.device', () => {
            expect( metadataClassObject.determineDeviceName( { device: 'metadata.device' } as cjson_metadata, {} as WebDriver.DesiredCapabilities ) ).toMatchSnapshot();
        } );

        it( 'should be able to return the device metadata based on the current.config.capabilities.deviceName', () => {
            expect(
                metadataClassObject.determineDeviceName( {} as cjson_metadata, { deviceName: 'current.config.capabilities.deviceName' } as WebDriver.DesiredCapabilities ),
            ).toMatchSnapshot();
        } );

        it( 'should be able to return the not known deviceName', () => {
            expect( metadataClassObject.determineDeviceName( {} as cjson_metadata, {} as WebDriver.DesiredCapabilities ) ).toMatchSnapshot();
        } );
    } );

    describe( 'determinePlatformName', () => {
        it( 'should be able to return the platform name based on the metadata.platform', () => {
            expect( metadataClassObject.determinePlatformName(
                {
                    platform: {
                        name: 'platform.name',
                    },
                } as cjson_metadata,
                {} as WebDriver.DesiredCapabilities,
            ) ).toMatchSnapshot();
        } );

        it( 'should be able to return the platform name based on the currentCapabilities.platformName', () => {
            expect(
                metadataClassObject.determinePlatformName( {} as cjson_metadata, { platformName: 'currentCapabilities.platformName' } as WebDriver.DesiredCapabilities ),
            ).toMatchSnapshot();
        } );

        it( 'should be able to return the platform name based on the currentCapabilities.platformName for mac properly', () => {
            expect(
                metadataClassObject.determinePlatformName( {} as cjson_metadata, { platformName: 'mac' } as WebDriver.DesiredCapabilities ),
            ).toMatchSnapshot();
        } );

        it( 'should be able to return the platform name based on the currentCapabilities.platformName for windows properly', () => {
            expect(
                metadataClassObject.determinePlatformName( {} as cjson_metadata, { platformName: 'windows nt' } as WebDriver.DesiredCapabilities ),
            ).toMatchSnapshot();
        } );

        it( 'should be able to return the not known platform name', () => {
            expect( metadataClassObject.determinePlatformName( {} as cjson_metadata, {} as WebDriver.DesiredCapabilities ) ).toMatchSnapshot();
        } );
    } );

    describe( 'determinePlatformVersion', () => {
        it( 'should be able to return the platform version based on the metadata.platform', () => {
            expect( metadataClassObject.determinePlatformVersion(
                {
                    platform: {
                        version: 'platform.version',
                    },
                } as cjson_metadata,
            ) ).toMatchSnapshot();
        } );

        it( 'should be able to return the not known platform version', () => {
            expect( metadataClassObject.determinePlatformVersion( {} as cjson_metadata ) ).toMatchSnapshot();
        } );
    } );

    describe( 'determineMetadata', () => {
        let determineAppDataSpy: jest.SpyInstance; let determineBrowserDataSpy: jest.SpyInstance; let determineDeviceNameSpy: jest.SpyInstance; let determinePlatformNameSpy: jest.SpyInstance;
        let determinePlatformVersionSpy;
        const appMockData = {
            app: {
                name: 'mock-appName',
                version: 'mock-appVersion',
            },
        };
        const browserMockData = {
            browser: {
                name: 'mock-browserName',
                version: 'mock-browserVersion',
            },
        };

        beforeEach( () => {
            delete global.browser;
            global.browser = {
                options: {
                    requestedCapabilities: {
                        w3cCaps: {
                            alwaysMatch: {
                                foo: true,
                            }
                        }
                    },
                } as WebdriverIOExtended,
            } as Browser<'sync'>;
            determineAppDataSpy = jest.spyOn( metadataClassObject, 'determineAppData' );
            determineBrowserDataSpy = jest.spyOn( metadataClassObject, 'determineBrowserData' ).mockReturnValue( browserMockData );
            determineDeviceNameSpy = jest.spyOn( metadataClassObject, 'determineDeviceName' ).mockReturnValue( NOT_KNOWN );
            determinePlatformNameSpy = jest.spyOn( metadataClassObject, 'determinePlatformName' ).mockReturnValue( NOT_KNOWN );
            determinePlatformVersionSpy = jest.spyOn( metadataClassObject, 'determinePlatformVersion' ).mockReturnValue( NOT_KNOWN );
        } );

        afterEach( () => {
            jest.clearAllMocks();
            delete global.browser;
        } );

        it( 'should return app metadata based on the currentCapabilities.app', () => {
            ( FULL_RUNNER_STATS.capabilities as W3CCapabilitiesExtended ).app = 'current.config.capabilities.app';

            determineAppDataSpy = jest.spyOn( metadataClassObject, 'determineAppData' ).mockReturnValue( appMockData );

            expect( metadataClassObject.determineMetadata( FULL_RUNNER_STATS ) ).toMatchSnapshot();

            expect( determineAppDataSpy ).toHaveBeenCalledTimes( 1 );
            expect( determineDeviceNameSpy ).toHaveBeenCalledTimes( 1 );
            expect( determinePlatformNameSpy ).toHaveBeenCalledTimes( 1 );
            expect( determinePlatformVersionSpy ).toHaveBeenCalledTimes( 1 );

            delete ( FULL_RUNNER_STATS.capabilities as W3CCapabilitiesExtended ).app;
            determineAppDataSpy.mockClear();
        } );

        it( 'should return app metadata based on the current.config.capabilities[\'cjson:metadata\'].app', () => {
            ( FULL_RUNNER_STATS.capabilities as W3CCapabilitiesExtended ).cjson_metadata.app = {
                'name': 'mock-appName',
                'version': 'mock-appVersion',
            };

            determineAppDataSpy = jest.spyOn( metadataClassObject, 'determineAppData' ).mockReturnValue( appMockData );

            expect( metadataClassObject.determineMetadata( FULL_RUNNER_STATS ) ).toMatchSnapshot();

            expect( determineAppDataSpy ).toHaveBeenCalledTimes( 1 );
            expect( determineDeviceNameSpy ).toHaveBeenCalledTimes( 1 );
            expect( determinePlatformNameSpy ).toHaveBeenCalledTimes( 1 );
            expect( determinePlatformVersionSpy ).toHaveBeenCalledTimes( 1 );

            delete ( FULL_RUNNER_STATS.capabilities as W3CCapabilitiesExtended ).cjson_metadata.app;
            determineAppDataSpy.mockClear();
        } );

        it( 'should return metadata based on the data.config.capabilities["cjson:metadata"]', () => {
            global.browser = {
                options: {
                    capabilities: {},
                } as Options.WebdriverIO,
            } as Browser<'sync'>;

            determineAppDataSpy = jest.spyOn( metadataClassObject, 'determineBrowserData' ).mockReturnValue( browserMockData );

            expect( metadataClassObject.determineMetadata( CAPS_METADATA_RUNNER_STATS ) ).toMatchSnapshot();
            determineAppDataSpy.mockClear();
        } );

        it( 'should return metadata based on the requestedCapabilities.w3cCaps.alwaysMatch', () => {
            global.browser = {
                options: {
                    capabilities: {},
                    requestedCapabilities: {
                        w3cCaps: {
                            alwaysMatch: {
                                'cjson_metadata': {},
                            },
                        },
                    },
                } as Options.WebdriverIO,
            } as Browser<'sync'>;

            determineAppDataSpy = jest.spyOn( metadataClassObject, 'determineBrowserData' ).mockReturnValue( browserMockData );

            expect( metadataClassObject.determineMetadata( WDIO6_RUNNER_STATS ) ).toMatchSnapshot();
            determineAppDataSpy.mockClear();
        } );

        it( 'should return metadata based on the browser.options.capabilities if w3cCaps is empty', () => {
            global.browser = {
                options: {
                    requestedCapabilities: {
                        w3cCaps: {

                        }
                    },
                    capabilities: {
                        browserName: 'chrome',
                        'cjson_metadata': {},
                    }
                } as WebdriverIOExtended,
            } as Browser<'sync'>;

            determineAppDataSpy = jest.spyOn( metadataClassObject, 'determineBrowserData' ).mockReturnValue( browserMockData );

            expect( metadataClassObject.determineMetadata( WDIO6_RUNNER_STATS ) ).toMatchSnapshot();
            determineAppDataSpy.mockClear();
        } );

        it( 'should return metadata based on the browser.options.capabilities', () => {
            global.browser = {
                options: {
                    capabilities: {
                        browserName: 'chrome',
                        'cjson_metadata': {},
                    },
                } as WebdriverIOExtended,

            } as Browser<'sync'>;

            determineAppDataSpy = jest.spyOn( metadataClassObject, 'determineBrowserData' ).mockReturnValue( browserMockData );

            expect( metadataClassObject.determineMetadata( WDIO6_RUNNER_STATS ) ).toMatchSnapshot();
            determineAppDataSpy.mockClear();
        } );

        it( 'should call determineBrowserData when there is no way to  determine the app data', () => {
            metadataClassObject.determineMetadata( SMALL_RUNNER_STATS );

            expect( determineAppDataSpy ).toHaveBeenCalledTimes( 0 );
            expect( determineBrowserDataSpy ).toHaveBeenCalledTimes( 1 );
            expect( determineDeviceNameSpy ).toHaveBeenCalledTimes( 1 );
            expect( determinePlatformNameSpy ).toHaveBeenCalledTimes( 1 );
            expect( determinePlatformVersionSpy ).toHaveBeenCalledTimes( 1 );
        } );

        it( 'should be able to return browser metadata', () => {
            expect( metadataClassObject.determineMetadata( SMALL_RUNNER_STATS ) ).toMatchSnapshot();

            expect( determineBrowserDataSpy ).toHaveBeenCalledTimes( 1 );
            expect( determineDeviceNameSpy ).toHaveBeenCalledTimes( 1 );
            expect( determinePlatformNameSpy ).toHaveBeenCalledTimes( 1 );
            expect( determinePlatformVersionSpy ).toHaveBeenCalledTimes( 1 );
        } );
    } );
} );
