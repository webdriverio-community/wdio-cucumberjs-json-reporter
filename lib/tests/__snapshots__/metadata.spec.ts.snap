// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metadata determineAppData should return that no app metadata could be determined 1`] = `
Object {
  "app": Object {
    "name": "No metadata.app.name available",
    "version": "No metadata.app.version available",
  },
}
`;

exports[`metadata determineAppData should return that the app name and version based on the metadata 1`] = `
Object {
  "app": Object {
    "name": "metadata app name",
    "version": "metadata version 1.2.3",
  },
}
`;

exports[`metadata determineAppData should return that the app name based on the config.app 1`] = `
Object {
  "app": Object {
    "name": "app.apk",
    "version": "No metadata.app.version available",
  },
}
`;

exports[`metadata determineBrowserData should return that no browser metadata could be determined 1`] = `
Object {
  "browser": Object {
    "name": "No metadata.browser.name available",
    "version": "No metadata.browser.version available",
  },
}
`;

exports[`metadata determineBrowserData should return that the browser name and version based on the capabilities when browserVersion is used 1`] = `
Object {
  "browser": Object {
    "name": "capabilities browser name",
    "version": "capabilities browserVersion",
  },
}
`;

exports[`metadata determineBrowserData should return that the browser name and version based on the capabilities when version is used 1`] = `
Object {
  "browser": Object {
    "name": "capabilities browser name",
    "version": "capabilities version",
  },
}
`;

exports[`metadata determineBrowserData should return that the browser name and version based on the configCapabilities 1`] = `
Object {
  "browser": Object {
    "name": "configCapabilities browser name",
    "version": "configCapabilities browser version",
  },
}
`;

exports[`metadata determineBrowserData should return that the browser name and version based on the metadata 1`] = `
Object {
  "browser": Object {
    "name": "metadata browser name",
    "version": "metadata version 1.2.3",
  },
}
`;

exports[`metadata determineDeviceName should be able to return the device metadata based on the current.config.capabilities.deviceName 1`] = `"current.config.capabilities.deviceName"`;

exports[`metadata determineDeviceName should be able to return the device metadata based on the metadata.device 1`] = `"metadata.device"`;

exports[`metadata determineDeviceName should be able to return the not known deviceName 1`] = `"Device name not known"`;

exports[`metadata determineMetadata should be able to return browser metadata 1`] = `
Object {
  "browser": Object {
    "name": "mock-browserName",
    "version": "mock-browserVersion",
  },
  "device": "not known",
  "platform": Object {
    "name": "not known",
    "version": "not known",
  },
}
`;

exports[`metadata determineMetadata should return app metadata based on the current.config.capabilities['cjson:metadata'].app 1`] = `
Object {
  "app": Object {
    "name": "mock-appName",
    "version": "mock-appVersion",
  },
  "device": "not known",
  "platform": Object {
    "name": "not known",
    "version": "not known",
  },
}
`;

exports[`metadata determineMetadata should return app metadata based on the currentCapabilities.app 1`] = `
Object {
  "app": Object {
    "name": "mock-appName",
    "version": "mock-appVersion",
  },
  "device": "not known",
  "platform": Object {
    "name": "not known",
    "version": "not known",
  },
}
`;

exports[`metadata determineMetadata should return metadata based on the browser.options.capabilities 1`] = `
Object {
  "browser": Object {
    "name": "mock-browserName",
    "version": "mock-browserVersion",
  },
  "device": "not known",
  "platform": Object {
    "name": "not known",
    "version": "not known",
  },
}
`;

exports[`metadata determineMetadata should return metadata based on the browser.options.capabilities if w3cCaps is empty 1`] = `
Object {
  "browser": Object {
    "name": "mock-browserName",
    "version": "mock-browserVersion",
  },
  "device": "not known",
  "platform": Object {
    "name": "not known",
    "version": "not known",
  },
}
`;

exports[`metadata determineMetadata should return metadata based on the data.config.capabilities["cjson:metadata"] 1`] = `
Object {
  "browser": Object {
    "name": "mock-browserName",
    "version": "mock-browserVersion",
  },
  "device": "not known",
  "platform": Object {
    "name": "not known",
    "version": "not known",
  },
}
`;

exports[`metadata determineMetadata should return metadata based on the requestedCapabilities.w3cCaps.alwaysMatch 1`] = `
Object {
  "browser": Object {
    "name": "mock-browserName",
    "version": "mock-browserVersion",
  },
  "device": "not known",
  "platform": Object {
    "name": "not known",
    "version": "not known",
  },
}
`;

exports[`metadata determinePlatformName should be able to return the not known platform name 1`] = `"Platform name not known"`;

exports[`metadata determinePlatformName should be able to return the platform name based on the currentCapabilities.platformName 1`] = `"currentCapabilities.platformName"`;

exports[`metadata determinePlatformName should be able to return the platform name based on the currentCapabilities.platformName for mac properly 1`] = `"osx"`;

exports[`metadata determinePlatformName should be able to return the platform name based on the currentCapabilities.platformName for windows properly 1`] = `"windows"`;

exports[`metadata determinePlatformName should be able to return the platform name based on the metadata.platform 1`] = `"platform.name"`;

exports[`metadata determinePlatformVersion should be able to return the not known platform version 1`] = `"Version not known"`;

exports[`metadata determinePlatformVersion should be able to return the platform version based on the metadata.platform 1`] = `"platform.version"`;

exports[`metadata determinePlatformVersion should be able to return the platform version based on the metadata.platform from desired capabilities 1`] = `"10.1"`;
